### PR TITLE
view-source command now using scheme

### LINKED ISSUE #2395: Use view-source scheme for :view-source
With QtWebEngine, we can use Chromium's `view-source:` scheme when using `:view-source`.

This gives us clickable links, and might even help with #1208 (when caret mode is implemented) and #1125.

### LINKED ISSUE #2948: When viewing source, status bar displays "Invalid URL!"
When opening the source of a webpage, the status bar screams Invalid URL! where there should be the current URL. The status bar should show the address normally, I think.

qutebrowser v0.11.0
Git commit: v0.11.0-393-g173688c74 (2017-08-14 22:11:12 +0200)
Backend: QtWebKit-NG (WebKit 602.1)

### LINKED ISSUE #3490: view-source: scheme doesn't load qutebrowser's internal JavaScript
As stated in the title, keybindings for scrolling - gg, G, ctrl+d and ctrl+u - don't work in view-source mode, though 'j' and 'k' work normally. I understand that most users don't scroll that much through source, but it would still be a nice thing to have, and doesn't seem much different from ordinary page scrolling.

***

qutebrowser v1.0.4
Git commit: 
Backend: QtWebEngine (Chromium 61.0.3163.140)

CPython: 3.6.4
Qt: 5.10.0
PyQt: 5.9.2

sip: 4.19.6
colorama: no
pypeg2: 2.15
jinja2: 2.10
pygments: 2.2.0
yaml: 3.12
cssutils: no
attr: 17.4.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.21.0
QtNetwork SSL: OpenSSL 1.1.0g  2 Nov 2017

Style: QFusionStyle
Platform: Linux-4.14.12-1-ARCH-x86_64-with-arch-Arch-Linux, 64bit
Linux distribution: Arch Linux (arch)
Frozen: False
Imported from /usr/lib/python3.6/site-packages/qutebrowser
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt

Paths:
cache: /home/lark/.cache/qutebrowser
config: /home/lark/.config/qutebrowser
data: /home/lark/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser
system data: /usr/share/qutebrowser

### PR DESCRIPTION
**Pull request by mjauvin**
Use WebEngine's view-source: scheme for "view-source" command.
Inject qutebrowser's JS files at DocumentReady
Also add missing URL when viewing source for WebKit.
    
Resolves #3490
Resolves #2395
Resolves #2948

A bit less intrusive than the last PR, but we still need to test this to make sure injecting Qutebrowser's JS files at DocumentReady instead of DocumentCreation does not cause problems elsewhere.

If you guys can try this on your end, I'll be running this code from now on and report any odd behavior.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3521)

### CODE REVIEW

**Comment by mjauvin** [comment]
@The-Compiler OK, I'll look into moving this to AbstractAction and the associated webengine/webkit classes.

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -758,7 +758,6 @@ def scroll_to_perc(self, perc: float = None, horizontal=False, count=None):
         else:
             x = None
             y = perc
-
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Unrelated change?

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -1512,23 +1511,7 @@ def view_source(self):
         except cmdexc.CommandError as e:
             message.error(str(e))
             return
-
-        def show_source_cb(source):
-            """Show source as soon as it's ready."""
-            # WORKAROUND for https://github.com/PyCQA/pylint/issues/491
-            # pylint: disable=no-member
-            lexer = pygments.lexers.HtmlLexer()
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: You should also remove the now-unused imports.

File: qutebrowser/browser/webengine/webenginetab.py [hunk 3]
```diff
@@ -99,6 +99,13 @@ def save_page(self):
         """Save the current page."""
         self._widget.triggerPageAction(QWebEnginePage.SavePage)
 
+    def show_source(self, dispatcher, url):
+        # use view-source: scheme to show page source for webengine
+        url = QUrl('view-source:{}'.format(url.toString()))
+        new_tab = dispatcher._tabbed_browser.tabopen(
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginetab.py]: Passing the `dispatcher` here creates a [coupling](https://en.wikipedia.org/wiki/Coupling_%28computer_programming%29) between two completely different parts of codes, which I'd rather avoid. Also, you're then accessing private variables on the dispatcher.

Instead, pass the `win_id` to the `*Action` objects when creating them, and get the tabbed-browser object using `tabbed_browser = objreg.get('tabbed-browser', scope='window', window=win_id)`.

File: qutebrowser/browser/webengine/webenginetab.py [hunk 4]
```diff
@@ -99,6 +99,13 @@ def save_page(self):
         """Save the current page."""
         self._widget.triggerPageAction(QWebEnginePage.SavePage)
 
+    def show_source(self, dispatcher, url):
+        # use view-source: scheme to show page source for webengine
+        url = QUrl('view-source:{}'.format(url.toString()))
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginetab.py]: Starting with Qt 5.8, there's the [QWebEnginePage.ViewSource](http://doc.qt.io/qt-5/qwebenginepage.html#WebAction-enum) action which should be used instead of opening the URL by hand.

Also, I'm not sure what formatting flags to pass to `toString` here, but I'm pretty sure the default doesn't do the right thing. You might want to check what happens with the WebAction with URLs with username/password, with URLs with UTF-8, and with URLs with special characters in it (like a `?query`).

File: qutebrowser/browser/webengine/webenginetab.py [hunk 5]
```diff
@@ -99,6 +99,13 @@ def save_page(self):
...
+        new_tab = dispatcher._tabbed_browser.tabopen(
+            url, background=True, related=True)
+        new_tab.data.viewing_source = True
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginetab.py]: If we set a `view-source:` URL for QtWebKit correctly, I actually think we don't need the `viewing_source` flag anymore as we can just check the URL. If that works out, you can remove it entirely.

File: qutebrowser/browser/webkit/webkittab.py [hunk 6]
```diff
@@ -50,6 +54,25 @@ def save_page(self):
...
+            formatter = pygments.formatters.HtmlFormatter(
+                full=True, linenos='table',
+                title='view-source:{}'.format(url.toDisplayString()))
+            # pylint: enable=no-member
+            highlighted = pygments.highlight(source, lexer, formatter)
+
+            new_tab = dispatcher._tabbed_browser.tabopen()
+            new_tab.set_html(highlighted)
+            new_tab.data.viewing_source = True
+            new_tab.url = lambda requested=False: QUrl(
```
Comment (The-Compiler) [inline qutebrowser/browser/webkit/webkittab.py]: You're [monkey patching](https://en.wikipedia.org/wiki/Monkey_patch) the `url` method away here. Apart from this being something you should never do outside of test code, it's also broken: No matter what you navigate this tab to in the future, it'll always show this URL, as you patched the original code away.

Instead, the `set_html` method can take a second argument which is the URL. Please use that instead.

File: qutebrowser/browser/webengine/webenginetab.py (hunk 3 above)
Comment (mjauvin) [reply qutebrowser/browser/webengine/webenginetab.py]: OK, nice!
What about the "dispatcher._current_widget()" in webkittab's show_source() method ?

Comment (mjauvin) [reply qutebrowser/browser/webengine/webenginetab.py]: Ahh, found the answer:

widget = self._tabbed_browser.currentWidget()

So I can use the currentWidget() method of the tabbed_browser I get with objreg... nice!

File: qutebrowser/browser/webengine/webenginetab.py (hunk 5 above)
Comment (mjauvin) [reply qutebrowser/browser/webengine/webenginetab.py]: You're right, I was able to get rid of the viewing_source flag. I also removed it from the TabData class.

File: qutebrowser/browser/browsertab.py [hunk 7]
```diff
@@ -139,6 +137,10 @@ def run_string(self, name):
             raise WebTabError("{} is not a valid web action!".format(name))
         self._widget.triggerPageAction(member)
 
+    def show_source(self, win_id, url):
```
Comment (The-Compiler) [inline qutebrowser/browser/browsertab.py]: You shouldn't pass the `win_id` and `url` here, as the tab knows both of those things itself, and this API should be as easy to use as possible (it's going to be used in a plugin API in the future).

Instead, pass the tab object when creating `AbstractAction` or its subclasses (see e.g. `AbstractCaret`) and then use `self._tab.win_id` and `self._tab.url()` or so.

File: qutebrowser/browser/webengine/webenginetab.py [hunk 8]
```diff
@@ -99,6 +99,19 @@ def save_page(self):
         """Save the current page."""
         self._widget.triggerPageAction(QWebEnginePage.SavePage)
 
+    def show_source(self, win_id, url):
+        try:
+            self._widget.triggerPageAction(QWebEnginePage.ViewSource)
+        except AttributeError:
+            # Qt < 5.8
+            # note: it's not possible to build the QUrl object by setting the
+            # scheme using setScheme('view-source').
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginetab.py]: Why not?

```
>>> u = QUrl()
>>> u.setScheme('view-source')
>>> u.setPath('http://www.example.com/')
>>> u
PyQt5.QtCore.QUrl('view-source:http://www.example.com/')
```

File: qutebrowser/browser/webkit/webkittab.py [hunk 9]
```diff
@@ -50,6 +54,34 @@ def save_page(self):
...
+            return s
+
+        def show_source_cb(source):
+            """Show source as soon as it's ready."""
+            # WORKAROUND for https://github.com/PyCQA/pylint/issues/491
+            # pylint: disable=no-member
+            lexer = pygments.lexers.HtmlLexer()
+            formatter = pygments.formatters.HtmlFormatter(
+                full=True, linenos='table',
+                title='view-source:' + format_url(url))
```
Comment (The-Compiler) [inline qutebrowser/browser/webkit/webkittab.py]: Why change the title to something both more complex and less readable?

File: qutebrowser/browser/webkit/webkittab.py [hunk 10]
```diff
@@ -50,6 +54,34 @@ def save_page(self):
...
+            # pylint: enable=no-member
+            highlighted = pygments.highlight(source, lexer, formatter)
+
+            base_url = QUrl('view-source:' + url.toString())
+            new_tab = tabbed_browser.tabopen(background=False, related=True)
+            new_tab.set_html(highlighted, base_url)
+
+        tabbed_browser = objreg.get('tabbed-browser', scope='window',
+                                    window=win_id)
+        tabbed_browser.currentWidget().dump_async(show_source_cb)
```
Comment (The-Compiler) [inline qutebrowser/browser/webkit/webkittab.py]: If you store the tab in `AbstractAction`, this is just `self._tab.dump_async(show_source_cb)`.

**Comment by The-Compiler** [comment]
Thanks!

**Comment by The-Compiler** [comment]
FYI: Only being able to use hints after a page fully loaded was quite annoying - so I did fac0f66e52c189344c8d5c27e0018c5909cda7f9 now, which just injects them at both points.
